File: tools/tap_device.c
#include <fcntl.h>
#include <linux/input.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/time.h>
#include <unistd.h>

static int emit(int fd, unsigned short type, unsigned short code, int value) {
    struct input_event event;
    memset(&event, 0, sizeof(event));
    gettimeofday(&event.time, NULL);
    event.type = type;
    event.code = code;
    event.value = value;
    return write(fd, &event, sizeof(event)) == sizeof(event) ? 0 : -1;
}

static int scale(int value, int screen_max, const struct input_absinfo *info) {
    return info->minimum + value * (info->maximum - info->minimum) / screen_max;
}

int main(int argc, char **argv) {
    int fd;
    int x;
    int y;
    struct input_absinfo x_info;
    struct input_absinfo y_info;

    if (argc != 3) {
        fprintf(stderr, "usage: tap_device X Y\n");
        return 2;
    }
    x = atoi(argv[1]);
    y = atoi(argv[2]);
    fd = open("/dev/input/event0", O_RDWR);
    if (fd < 0 || ioctl(fd, EVIOCGABS(ABS_MT_POSITION_X), &x_info) < 0 ||
        ioctl(fd, EVIOCGABS(ABS_MT_POSITION_Y), &y_info) < 0) {
        perror("touchscreen");
        return 1;
    }
    x = scale(x, 1023, &x_info);
    y = scale(y, 599, &y_info);

    emit(fd, EV_ABS, ABS_MT_SLOT, 0);
    emit(fd, EV_ABS, ABS_MT_TRACKING_ID, 123);
    emit(fd, EV_ABS, ABS_MT_TOUCH_MAJOR, 20);
    emit(fd, EV_ABS, ABS_MT_POSITION_X, x);
    emit(fd, EV_ABS, ABS_MT_POSITION_Y, y);
    emit(fd, EV_KEY, BTN_TOUCH, 1);
    emit(fd, EV_SYN, SYN_REPORT, 0);
    usleep(80000);
    emit(fd, EV_KEY, BTN_TOUCH, 0);
    emit(fd, EV_ABS, ABS_MT_TRACKING_ID, -1);
    emit(fd, EV_SYN, SYN_REPORT, 0);
    close(fd);
    return 0;
}

File: tools/tap_device.py
#!/usr/bin/env python3
"""Inject one protocol-B touchscreen tap through /dev/input/event0."""

from __future__ import annotations

import argparse
import struct
import subprocess
import time
from pathlib import Path


ADB = Path(r"D:\Install\AndroidPlatformTools\adb.exe")
SERIAL = "192.168.1.70:5555"
INPUT_EVENT = "/dev/input/event0"

EV_SYN = 0
EV_KEY = 1
EV_ABS = 3
SYN_REPORT = 0
BTN_TOUCH = 330
ABS_MT_SLOT = 0x2F
ABS_MT_POSITION_X = 0x35
ABS_MT_POSITION_Y = 0x36
ABS_MT_TRACKING_ID = 0x39


def event(event_type: int, code: int, value: int) -> bytes:
    # Target userspace is 32-bit: timeval (2 x int32), type/code, value.
    return struct.pack("<IIHHi", 0, 0, event_type, code, value)


def report(events: list[tuple[int, int, int]]) -> bytes:
    return b"".join(event(*item) for item in events) + event(EV_SYN, SYN_REPORT, 0)


def write_events(payload: bytes) -> None:
    subprocess.run(
        [str(ADB), "-s", SERIAL, "shell", "dd", f"of={INPUT_EVENT}"],
        input=payload,
        check=True,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("x", type=int)
    parser.add_argument("y", type=int)
    args = parser.parse_args()

    subprocess.run(
        [
            str(ADB),
            "-s",
            SERIAL,
            "shell",
            "/tmp/tap_device",
            str(args.x),
            str(args.y),
        ],
        check=True,
    )


if __name__ == "__main__":
    main()
